File: ABC/abc347/d.cpp
#include <bits/stdc++.h>

using namespace std;


// Sとgood_string間のコストを計算する関数
long long calculateCost(const string& S, long long good_string, const vector<int>& C, int N) {
    long long xorValue = stoll(S, nullptr, 2) ^ good_string;
    long long cost = 0;
    for (int i = 0; i < N; ++i) {
        if (xorValue & (1LL << i)) {
            cost += C[i];
        }
    }
    return cost;
}

// すべての良い文字列を生成する関数
void generateGoodStrings(int N, const string& s, vector<long long>& goodStrings) {
    if (s.size() == N) {
        if (countAdjacent(s)) {
            goodStrings.push_back(stoll(s, nullptr, 2));
        }
        return;
    }
    generateGoodStrings(N, s + '0', goodStrings);
    generateGoodStrings(N, s + '1', goodStrings);
}

// 連続する同じ文字が1つだけあるかどうかを確認する関数
bool countAdjacent(const string& s) {
    int count = 0;
    for (size_t i = 0; i < s.size() - 1; ++i) {
        if (s[i] == s[i + 1]) {
            count++;
        }
    }
    return count == 1;
}

int main() {
    int N;
    cin >> N;

    string S;
    cin >> S;

    vector<int> C(N);
    for (int i = 0; i < N; ++i) {
        cin >> C[i];
    }

    // コスト配列を反転させる
    reverse(C.begin(), C.end());

    vector<long long> goodStrings;
    generateGoodStrings(N, "", goodStrings);

    long long minCost = LLONG_MAX;

    for (long long goodString : goodStrings) {
        minCost = min(minCost, calculateCost(S, goodString, C, N));
    }

    cout << minCost << endl;

    return 0;
}

File: ABC/abc347/d.py

N = int(input())

S = input()

C = list(map(int, input().split()))


# def generate_good_strings(N):

#     # 長さNのすべての良い文字列を生成する関数
#     def generate_all_strings(N, s, result):
#         if len(s) == N:
#             result.append(s)
#             return
#         generate_all_strings(N, s + '0', result)
#         generate_all_strings(N, s + '1', result)

#     # 2つの連続する文字が同じかどうかをチェックする関数
#     def is_good_string(s):
#         count = 0
#         for i in range(len(s) - 1):
#             if s[i] == s[i + 1]:
#                 count += 1
#         return count == 1

#     all_strings = []
#     good_strings = []

#     generate_all_strings(N, '', all_strings)

#     for s in all_strings:
#         if is_good_string(s):
#             good_strings.append(int(s, 2)) # 2進数を10進数に変換して追加
    
#     return good_strings

### 効率良く良い文字列を生成する ###
# good_strings = []

# # N-1文字の101010..or 010101..を生成する
# string_start_0 = '0' + '10' * (N//2)
# string_start_1 = '1' + '01' * (N//2)

# # どこに0または1を入れるかを考える
# for i in range(N):
#     good_strings.append(int(string_start_0[:N], 2))

def generate_good_strings(N):
    good_strings = []

    # パターン1: 先頭に連続する2つの同じ文字、残りは異なる文字
    for i in range(2, N + 1):
        good_strings.append('1' * i + '0' * (N - i))  # 先頭が1
        good_strings.append('0' * i + '1' * (N - i))  # 先頭が0

    # パターン2: 最初の文字と異なる2文字が連続し、残りは最初の文字と同じ
    if N > 2:
        for i in range(1, N - 1):
            good_strings.append('1' * i + '01' + '1' * (N - i - 2))  # 先頭が1
            good_strings.append('0' * i + '10' + '0' * (N - i - 2))  # 先頭が0

    # 二進数を十進数に変換
    return [int(s, 2) for s in good_strings]

good_strings = generate_good_strings(N)

print(good_strings)

min_cost = float('inf')

S = int(S, 2)

C = C[::-1]

for good_string in good_strings:

    # Sとの排他的論理和を取って，コストを計算する
    xor = S ^ good_string

    cost = 0
    for i in range(N):
        if xor & (1 << i):
            cost += C[i]

    min_cost = min(min_cost, cost)
    
print(min_cost)






